enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,906 @@
+* openshift/openshift-apiserver
+
+The router configures HAProxy using a configuration template. The template is populated with data from `ServiceAliasConfig` values, where each `ServiceAliasConfig` represents a Route.
+The enhancement adds four fields in `ServiceAliasConfig`: `SetHTTPResponseHeaders`, `DeleteHTTPResponseHeaders`, `SetHTTPRequestHeaders`, and `DeleteHTTPRequestHeaders` for all combinations of the "Set" and "Delete" actions and request or response headers to be set or deleted in the
+the `haproxy.config` sections corresponding to the following backends: plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption ([code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609)).
+The router sets this variable based on the Route's `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` field values. `ServiceAliasConfig` is populated by [the router's `createServiceAliasConfig` method](https://github.com/openshift/router/blob/3065f6583f3925328fbdbfe95e3bc7bb7a084d33/pkg/router/template/router.go#L908-L935)
+
+```go
+// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
+type ServiceAliasConfig struct {
+    ...
+	// For route based setting of http response headers
+	SetHTTPResponseHeaders              []SetHTTPHeader
+	DeleteHTTPResponseHeaders           []DeleteHTTPHeader
+	// For route based setting of http request headers
+	SetHTTPRequestHeaders              []SetHTTPHeader
+	DeleteHTTPRequestHeaders           []DeleteHTTPHeader
+}
+```
+```go
+// SetHTTPHeader specifies an HTTP header that should be set.
+type SetHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+
+	// Value specifies the header value.
+	Value string
+
+	// Action specifies the action to be performed.
+	Action string
+}
+
+type DeleteHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+}
+```
+
+In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside router. So, we will run a loop
+on either `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` which will have the desired
+sequence.
+
+C) library-go code will be having validations for checking Route API fields.
+   We will be creating an immutable/hard-coded deny-list of HTTP response headers in [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141)
+   which won't be allowed to be set or deleted. For the initial enhancement, the HSTS `strict-transport-security` header name is included in this deny-list.
```

Comment:
More details needed here about any other headers that should be in the deny-list, or else an explanation as to why it was determined that we 1) wouldn't notify users when a header wasn't allowed and 2) would allow users to override values we have explicitly hard-coded in the HAProxy template.

Reply:
> More details needed here about any other headers that should be in the deny-list, or else an explanation as to why it was determined that we 1) wouldn't notify users when a header wasn't allowed 

The API will throw an message if HSTS is added.

and 2) would allow users to override values we have explicitly hard-coded in the HAProxy template.


 The route owner does over-ride the headers set by admin however the scope for it will be at his application level only and it won't cause any issues cluster-wide as we don't set these in the front-end. However cluster-admin should carefully set the headers which get set in the front end sections for all routes. We can may be document a list of request headers not permitted to be set by  the admin.

Did you get a chance to look at this https://github.com/openshift/router/pull/438#discussion_r1170637633 ?

Reply:
as discussed the EP has been updated with the relevant documentation in the Non-goals section